csrc/fusion_profiler.cpp

```diff
@@ -595,6 +462,185 @@ std::ostream& operator<<(std::ostream& os, const FusionProfile& fp) {
+  CUptiResult status = CUPTI_SUCCESS;
+
+  // This is an arbitrary record limit to make sure we do not get into an
+  // infinite loop;
+  const size_t max_records = 10000;
+  bool found_max_limit = false;
+
+  for (size_t i = 0; i < max_records; ++i) {
+    status = cuptiActivityGetNextRecord(pBuffer, validBytes, &pRecord);
+    if (status == CUPTI_SUCCESS) {
+      // Processes a valid CUPTI Activty record and records it with the
+      // fusion profiling infrastructure if the record is of interest.
+      record_cupti_activity(pRecord, pFileHandle);
+    } else if (status == CUPTI_ERROR_MAX_LIMIT_REACHED) {
+      // This case is hit every time an activity buffer is read and you reach
+      // the end of the buffer activity to consume. Therefore, it is the
+      // intended way to break out of the loop!  The name is miss leading as
+      // you aren't reaching a Max Limit, necessarily!
+      found_max_limit = true;
+      break;
+    } else {
+      NVFUSER_CUPTI_SAFE_CALL(status);
+    }
+  }
+  NVF_ERROR(
+      found_max_limit,
+      "The CUPTI buffer has more than ",
+      max_records,
+      " record! Is that expected?");
+}
+
+// The functions cupti_buffer_requested and cupti_buffer_completed are
+// registered with the CUPTI Activiy Record Callback API:
+// cuptiActivityRegisterCallbacks.  Each of the functions APIs is prescribed
+// by CUPTI and you can find their signatured definitions in the CUPT docs.
+
+void cupti_buffer_requested(
+    uint8_t** ppBuffer,
+    size_t* pSize,
+    size_t* pMaxNumRecords) {
+  uint8_t* pBuffer = FusionProfiler::cuptiBufferPtr();
+  NVF_ERROR(pBuffer, "CUPTI Activity Record buffer pointer is null!");
+  const size_t align_size = 8;
+  NVF_ERROR(
+      ((uintptr_t)pBuffer & (align_size - 1)) == 0,
+      "The CUPTI Activity Record buffer needs to be 8 byte aligned!");
+
+  *ppBuffer = pBuffer;
+  *pSize = FusionProfiler::cupti_activity_buffer_size;
+  // NOTE: The Max Number of records limits the number of records that can be
+  // recorded in the activity buffer.  When set to 0, it puts as many records
+  // as it can which effectively disables a max limit.
+  *pMaxNumRecords = 0;
+}
+
+void cupti_buffer_completed(
+    CUcontext context,
+    uint32_t streamId,
+    uint8_t* pBuffer,
+    size_t size,
+    size_t validSize) {
+  if (validSize > 0) {
+    record_cupti_activity_buffer(pBuffer, validSize, stdout, nullptr);
+  }
+}
+
+// CUPT activities to enable/disable
+constexpr CUpti_ActivityKind cupti_activities[] = {
+    CUPTI_ACTIVITY_KIND_CONCURRENT_KERNEL,
+    CUPTI_ACTIVITY_KIND_DRIVER,
+    CUPTI_ACTIVITY_KIND_RUNTIME,
+    CUPTI_ACTIVITY_KIND_EXTERNAL_CORRELATION};
+
+void enableCuptiActivities() {
+  for (const auto& activity : cupti_activities) {
+    NVFUSER_CUPTI_SAFE_CALL(cuptiActivityEnable(activity));
+  }
+}
+
+void disableCuptiActivities() {
+  for (const auto& activity : cupti_activities) {
+    NVFUSER_CUPTI_SAFE_CALL(cuptiActivityDisable(activity));
+  }
```

Comment:
nitpick: `cuptiActivityFlushAll` seems to always follow disableCuptiActivities. should we move it inside and rename this?

Reply:
I prefer keeping them separate, since they are fundamentally not tied together.
The flag for `cuptiActivityFlushAll (0/1)` also differs when we want to read the buffer, or forcefully clear it.